File: pkg/store/notification_log_store.go
package store

import (
	"context"
	"encoding/csv"
	"encoding/json"
	"fmt"
	"io"
	"time"

	"gorm.io/gorm"

	"github.com/PineappleBond/xyncra-server/pkg/store/model"
)

// NotificationLogStore provides data access operations for push notification
// logging and deduplication.
type NotificationLogStore struct {
	db *gorm.DB
}

// Save inserts a new notification log record.
func (ns *NotificationLogStore) Save(ctx context.Context, log *model.NotificationLog) error {
	if err := ns.db.WithContext(ctx).Create(log).Error; err != nil {
		return classifyError(fmt.Errorf("store: save notification log: %w", err))
	}
	return nil
}

// NotificationLogFilter defines optional filters for listing notification logs.
type NotificationLogFilter struct {
	StartTime *time.Time
	EndTime   *time.Time
	Type      string
	Limit     int
}

// List returns notification logs matching the given filters, ordered by
// CreatedAt descending (newest first).
func (ns *NotificationLogStore) List(ctx context.Context, filter NotificationLogFilter) ([]*model.NotificationLog, error) {
	query := ns.db.WithContext(ctx).Model(&model.NotificationLog{})

	if filter.StartTime != nil {
		query = query.Where("created_at >= ?", *filter.StartTime)
	}
	if filter.EndTime != nil {
		query = query.Where("created_at <= ?", *filter.EndTime)
	}
	if filter.Type != "" {
		query = query.Where("type = ?", filter.Type)
	}

	limit := filter.Limit
	if limit <= 0 || limit > 1000 {
		limit = 100
	}

	var logs []*model.NotificationLog
	err := query.
		Order("created_at DESC").
		Limit(limit).
		Find(&logs).Error
	if err != nil {
		return nil, classifyError(fmt.Errorf("store: list notification logs: %w", err))
	}
	return logs, nil
}

// ListBySeqRange returns notification logs with Seq in the range
// [startSeq, endSeq] (inclusive), ordered by Seq ascending.
func (ns *NotificationLogStore) ListBySeqRange(ctx context.Context, startSeq, endSeq uint32) ([]*model.NotificationLog, error) {
	var logs []*model.NotificationLog
	err := ns.db.WithContext(ctx).
		Where("seq >= ? AND seq <= ?", startSeq, endSeq).
		Order("seq ASC").
		Find(&logs).Error
	if err != nil {
		return nil, classifyError(fmt.Errorf("store: list notification logs by seq range: %w", err))
	}
	return logs, nil
}

// ExportCSV writes notification logs matching the filter as CSV to the given writer.
func (ns *NotificationLogStore) ExportCSV(ctx context.Context, w io.Writer, filter NotificationLogFilter) error {
	logs, err := ns.List(ctx, filter)
	if err != nil {
		return err
	}

	cw := csv.NewWriter(w)
	defer cw.Flush()

	if err := cw.Write([]string{"id", "seq", "type", "created_at"}); err != nil {
		return fmt.Errorf("store: write csv header: %w", err)
	}

	for _, l := range logs {
		record := []string{
			l.ID,
			fmt.Sprintf("%d", l.Seq),
			l.Type,
			l.CreatedAt.Format(time.RFC3339),
		}
		if err := cw.Write(record); err != nil {
			return fmt.Errorf("store: write csv record: %w", err)
		}
	}
	return nil
}

// ExportJSON writes notification logs matching the filter as JSON to the given writer.
func (ns *NotificationLogStore) ExportJSON(ctx context.Context, w io.Writer, filter NotificationLogFilter) error {
	logs, err := ns.List(ctx, filter)
	if err != nil {
		return err
	}

	enc := json.NewEncoder(w)
	enc.SetIndent("", "  ")
	if err := enc.Encode(logs); err != nil {
		return fmt.Errorf("store: encode notification logs json: %w", err)
	}
	return nil
}

// CleanupBefore hard-deletes notification logs with CreatedAt strictly before
// the given time. Returns the number of deleted rows.
func (ns *NotificationLogStore) CleanupBefore(ctx context.Context, before time.Time) (int64, error) {
	result := ns.db.WithContext(ctx).
		Unscoped().
		Where("created_at < ?", before).
		Delete(&model.NotificationLog{})
	if result.Error != nil {
		return 0, classifyError(fmt.Errorf("store: cleanup notification logs: %w", result.Error))
	}
	return result.RowsAffected, nil
}

// GetLatestSeq returns the highest Seq value in the notification log.
// Returns 0 if the log is empty.
func (ns *NotificationLogStore) GetLatestSeq(ctx context.Context) (uint32, error) {
	var seq uint32
	err := ns.db.WithContext(ctx).
		Model(&model.NotificationLog{}).
		Select("COALESCE(MAX(seq), 0)").
		Scan(&seq).Error
	if err != nil {
		return 0, classifyError(fmt.Errorf("store: get latest notification seq: %w", err))
	}
	return seq, nil
}

File: demo/web/packages/xyncra-client-core/src/db/notification-log-store.ts
/**
 * NotificationLogStore — data access operations for push notification logging.
 *
 * Mirrors Go NotificationLogStore (pkg/store/notification_log_store.go).
 *
 * Key constraint:
 *   C6 — NotificationLog.Seq has a unique index for deduplication.
 *        The save() method catches ConstraintError and throws ErrDuplicateKey.
 */

import type Dexie from 'dexie';
import type { Transaction } from 'dexie';

import { ErrDuplicateKey, ErrNotFound } from '../errors';
import type { XyncraDatabase } from './index';
import type { NotificationLog } from './models';

// ---------------------------------------------------------------------------
// Filter types
// ---------------------------------------------------------------------------

/**
 * NotificationLogFilter defines optional filters for listing notification logs.
 * Mirrors Go store.NotificationLogFilter.
 */
export interface NotificationLogFilter {
  start_time?: Date;
  end_time?: Date;
  type?: string;
  limit?: number;
}

// ---------------------------------------------------------------------------
// Store
// ---------------------------------------------------------------------------

/**
 * NotificationLogStore provides data access operations for push notification
 * logging and deduplication.
 */
export class NotificationLogStore {
  constructor(private readonly db: XyncraDatabase) {}

  /**
   * Inserts a new notification log record.
   * Throws ErrDuplicateKey if a record with the same seq already exists (C6).
   */
  async save(log: NotificationLog): Promise<void> {
    try {
      await this.db.notificationLogs.add(log);
    } catch (error) {
      if (this.isConstraintError(error)) {
        throw ErrDuplicateKey;
      }
      throw error;
    }
  }

  /**
   * Returns notification logs matching the given filters, ordered by created_at
   * descending (newest first).
   */
  async list(filter: NotificationLogFilter): Promise<NotificationLog[]> {
    let limit = filter.limit ?? 100;
    if (limit <= 0 || limit > 1000) limit = 100;

    let logs: NotificationLog[];

    if (filter.type) {
      logs = await this.db.notificationLogs
        .where('type')
        .equals(filter.type)
        .toArray();
    } else {
      logs = await this.db.notificationLogs.toArray();
    }

    // Apply time range filters in JS.
    if (filter.start_time) {
      const startMs = filter.start_time.getTime();
      logs = logs.filter((l) => l.created_at.getTime() >= startMs);
    }
    if (filter.end_time) {
      const endMs = filter.end_time.getTime();
      logs = logs.filter((l) => l.created_at.getTime() <= endMs);
    }

    // Sort by created_at descending.
    logs.sort(
      (a, b) =>
        new Date(b.created_at).getTime() - new Date(a.created_at).getTime(),
    );

    return logs.slice(0, limit);
  }

  /**
   * Returns notification logs with seq in the range [startSeq, endSeq]
   * (inclusive), ordered by seq ascending.
   */
  async listBySeqRange(
    startSeq: number,
    endSeq: number,
  ): Promise<NotificationLog[]> {
    const logs = await this.db.notificationLogs
      .where('seq')
      .between(startSeq, endSeq, true, true)
      .toArray();

    // Sort by seq ascending.
    logs.sort((a, b) => a.seq - b.seq);

    return logs;
  }

  /**
   * Exports notification logs matching the filter as CSV string.
   */
  async exportCSV(filter: NotificationLogFilter): Promise<string> {
    const logs = await this.list(filter);

    const header = 'id,seq,type,created_at';
    const rows = logs.map((l) => {
      return [l.id, String(l.seq), l.type, l.created_at.toISOString()].join(
        ',',
      );
    });

    return [header, ...rows].join('\n');
  }

  /**
   * Exports notification logs matching the filter as JSON string.
   */
  async exportJSON(filter: NotificationLogFilter): Promise<string> {
    const logs = await this.list(filter);
    return JSON.stringify(logs, null, 2);
  }

  /**
   * Hard-deletes notification logs with created_at strictly before the given time.
   * Returns the number of deleted rows.
   */
  async cleanupBefore(before: Date): Promise<number> {
    return this.db.notificationLogs.where('created_at').below(before).delete();
  }

  /**
   * Returns the number of notification logs with created_at strictly before the
   * given time without deleting them.
   */
  async countBefore(before: Date): Promise<number> {
    return this.db.notificationLogs.where('created_at').below(before).count();
  }

  /**
   * Returns the highest seq value in the notification log.
   * Returns 0 if the log is empty.
   */
  async getLatestSeq(): Promise<number> {
    // seq is the primary key; lastKey() returns the highest key without
    // loading all records.
    const maxSeq = await this.db.notificationLogs.orderBy('seq').lastKey();
    return (maxSeq as number) ?? 0;
  }

  /**
   * Inserts a notification log record within the given transaction.
   * Throws ErrDuplicateKey if a record with the same seq already exists (C6).
   */
  async saveTx(tx: Transaction, log: NotificationLog): Promise<void> {
    const table = tx.table('notificationLogs') as Dexie.Table<
      NotificationLog,
      number
    >;
    try {
      await table.add(log);
    } catch (error) {
      if (this.isConstraintError(error)) {
        throw ErrDuplicateKey;
      }
      throw error;
    }
  }

  /** Checks if an error is a constraint violation. */
  private isConstraintError(error: unknown): boolean {
    if (error instanceof Error) {
      return error.name === 'ConstraintError';
    }
    return false;
  }
}
